lib/AST/ASTImporter.cpp

```diff
@@ -1028,18 +1028,40 @@ void ASTNodeImporter::ImportDeclContext(DeclContext *FromDC, bool ForceImport,
     Importer.ImportContext(FromDC);
     return;
   }
-  llvm::SmallVector<Decl*, 8> Decls;
-  for (auto *From : FromDC->decls())
-    Decls.push_back(Importer.Import(From));
-  if (ToDC) {
-    // Restore the order.
-    for (auto *D : Decls) {
-      if (D && D->getLexicalDeclContext() == ToDC)
-        ToDC->removeDecl(D);
+  llvm::SmallVector<Decl*, 8> ImportedDecls;
+  for (Decl *From : FromDC->decls())
+    ImportedDecls.push_back(Importer.Import(From));
+
+  if (ToDC) { // Restore the order. O(n*n) in worst case, because the remove.
+
+    llvm::SmallDenseSet<Decl *, 2> MissingDecls;
+
+    // Frist, remove all declarations, which may be in wrong order.
+    for (Decl *ToD : ImportedDecls) {       // O(n)
+      if (ToD && ToDC->containsDecl(ToD)) { // containsDecl is O(1)
```

Comment:
I would leave a comment that we only need to reorder decls in the LexicalDeclContext.